File: programmers/코딩테스트 고득점 Kit/해시/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include<algorithm>
#include<iostream>
using namespace std;
map<string, pair<int, int>> m;
int cnt, chk[105];
vector<pair<int, string>> sv;
vector<pair<int, pair<string, int>>> splays;
vector<int> solution(vector<string> genres, vector<int> plays) {
	vector<int> answer;
	for (int i = 0; i < plays.size(); i++) {
		if (m[genres[i]].first == 0) m[genres[i]].first = ++cnt;
		m[genres[i]].second += plays[i];
		splays.push_back({ plays[i],{genres[i],100000 - i} });
	}
	for (int i = 0; i < plays.size(); i++) {
		if (chk[m[genres[i]].first] == 0) {
			chk[m[genres[i]].first] = 1;
			sv.push_back({ m[genres[i]].second, genres[i] });
		}
	}
	sort(sv.begin(), sv.end());
	sort(splays.begin(), splays.end());
	for (int i = sv.size() - 1; i >= 0; i--) {
		int count = 0;
		for (int j = splays.size() - 1; j >= 0; j--) {
			if (count == 2) break;
			if (sv[i].second == splays[j].second.first) {
				count++;
				answer.push_back(100000 - splays[j].second.second);
			}
		}
	}
	return answer;
}

File: programmers/코딩테스트 고득점 Kit/해시/베스트앨범.java
import java.util.*;
class Album {
    private int playCount;
    private int genre;
    private int uniqueNumber;
    Album(int playCount, int genre, int uniqueNumber) {
        this.playCount = playCount;
        this.genre = genre;
        this.uniqueNumber = uniqueNumber;
    }

    public int getPlayCount() {
        return playCount;
    }

    public int getGenre() {
        return genre;
    }

    public int getUniqueNumber() {
        return uniqueNumber;
    }
}

class Genres {
    private String name;
    private int playCount;

    public String getName() {
        return name;
    }

    public int getPlayCount() {
        return playCount;
    }

    public Genres(String name, int playCount) {
        this.name = name;
        this.playCount = playCount;
    }
}

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Album> albumList = new ArrayList<>();
        Map<String, Integer> genrePlaySum = new HashMap<>();
        Map<String, Integer> genreRank = new HashMap<>();
        Map<Integer, Integer> genreCountInBestAlbum = new HashMap<>();

        //장르별 플레이수 합계
        for(int i=0;i<genres.length;i++) {
            if(genrePlaySum.containsKey(genres[i])) genrePlaySum.put(genres[i], genrePlaySum.get(genres[i]) + plays[i]);
            else genrePlaySum.put(genres[i], plays[i]);
            genreRank.put(genres[i], 0);
        }

        //장르별 순위 계산
        List<Genres> genresList = new ArrayList<>();
        for (String genre : genreRank.keySet()) {
            genresList.add(new Genres(genre, genrePlaySum.get(genre)));
        }
        genresList.sort(Comparator.comparing(Genres::getPlayCount).reversed());
        for(int i=0;i<genresList.size();i++) {
            genreRank.put(genresList.get(i).getName(), i+1);
            genreCountInBestAlbum.put(i+1, 0);
        }

        //리스트 만들기
        for(int i=0;i<genres.length;i++)
            albumList.add(new Album(10000-plays[i], genreRank.get(genres[i]), i));
        ArrayList<Integer> arrayList = new ArrayList<>();
        albumList.sort(Comparator.comparing(Album::getGenre).thenComparing(Album::getPlayCount).thenComparing(Album::getUniqueNumber));

        for(Album album : albumList) {
            if (genreCountInBestAlbum.get(album.getGenre()) == 2) continue;
            genreCountInBestAlbum.put(album.getGenre(), genreCountInBestAlbum.get(album.getGenre()) + 1);
            arrayList.add(album.getUniqueNumber());
        }
        int[] answer = new int[arrayList.size()];
        for(int i=0;i< arrayList.size();i++) {
            answer[i] = arrayList.get(i);
        }
        return answer;
    }
}